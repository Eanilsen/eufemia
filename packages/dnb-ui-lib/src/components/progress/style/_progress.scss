/*
 * Progress complightnt
 *
 */

:root {
  --progress-circular-circle: 88;
  --progress-circular-circle-offset--min: -1;
  --progress-circular-circle-offset--max: -88;
}

.dnb-progress {
  position: relative;

  // circular variant
  &__circular {
    position: relative;

    // medium
    width: 2rem;
    height: 2rem;

    &--small {
      width: 1rem;
      height: 1rem;
    }

    &--large {
      width: 4rem;
      height: 4rem;
    }

    &--huge {
      width: 20rem;
      height: 20rem;
    }

    // since SVG is starting 90deg from top
    transform: rotate(-90deg);

    svg {
      position: absolute;
      width: 100%;
      height: 100%;
    }

    &__line {
      animation-duration: 2s;
      animation-delay: 200ms;
      animation-timing-function: linear;
      animation-iteration-count: infinite;
    }

    &__line.light {
      // z-index: 2;
      animation-name: progress-circular-line-light;
      stroke-dasharray: var(--progress-circular-circle),
        var(--progress-circular-circle);
      stroke-dashoffset: var(--progress-circular-circle-offset--min);
    }
    &__line.dark {
      // z-index: 3;
      animation-name: progress-circular-line-dark;
      stroke-dasharray: var(--progress-circular-circle),
        var(--progress-circular-circle);
      stroke-dashoffset: var(--progress-circular-circle-offset--min);
    }
    &__line.paused {
      // z-index: 1;
      animation-play-state: paused;
    }

    &__circle {
      stroke-linecap: round;
    }
    &__line.light &__circle {
      stroke: grey;
    }
    &__line.dark &__circle {
      stroke: black;
    }
  }
}

@keyframes progress-circular-line-light {
  0% {
    stroke-dasharray: var(--progress-circular-circle),
      var(--progress-circular-circle);
    stroke-dashoffset: var(--progress-circular-circle-offset--min);
  }
  50% {
    stroke-dasharray: var(--progress-circular-circle),
      var(--progress-circular-circle);
    stroke-dashoffset: var(--progress-circular-circle-offset--min);
  }
  100% {
    stroke-dasharray: var(--progress-circular-circle),
      var(--progress-circular-circle);
    stroke-dashoffset: var(--progress-circular-circle-offset--max);
  }
}

@keyframes progress-circular-line-dark {
  0% {
    stroke-dasharray: var(--progress-circular-circle),
      var(--progress-circular-circle);
    stroke-dashoffset: var(--progress-circular-circle-offset--min);
  }
  50% {
    stroke-dasharray: var(--progress-circular-circle),
      var(--progress-circular-circle);
    stroke-dashoffset: var(--progress-circular-circle-offset--max);
  }
  100% {
    stroke-dasharray: var(--progress-circular-circle),
      var(--progress-circular-circle);
    stroke-dashoffset: var(--progress-circular-circle-offset--max);
  }
}
