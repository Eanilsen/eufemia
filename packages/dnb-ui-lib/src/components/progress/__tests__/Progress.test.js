/**
 * Component Test
 *
 */

import React from 'react'
import {
  mount,
  fakeProps,
  axeComponent,
  toJson,
  loadScss
} from '../../../core/jest/jestSetup'
import Component from '../Progress'

// just to make sure we re-run the test in watch mode due to changes in theese files
import _progress from '../style/_progress.scss' // eslint-disable-line
import dnb_progress from '../style/dnb-progress.scss' // eslint-disable-line
import dnb_progress_theme_ui from '../style/themes/dnb-progress-theme-ui.scss' // eslint-disable-line

const props = fakeProps(require.resolve('../Progress'), {
  all: true,
  optional: true
})

describe('Circular Progress component', () => {
  const Comp = mount(
    <Component {...props} type="circular" progress={50} />
  )

  it('have to match snapshot', () => {
    expect(toJson(Comp)).toMatchSnapshot()
  })

  it('has to have a stroke-dashoffset of -44 on 50%', () => {
    expect(
      Comp.find('svg.dnb-progress__circular__line.light[style]')
        .instance()
        .getAttribute('style')
    ).toBe('stroke-dashoffset: -44;')
  })

  it('has to have a aria-label with a 50% value', () => {
    expect(
      Comp.find('.dnb-progress__circular')
        .instance()
        .getAttribute('aria-label')
    ).toBe('50%')
  })

  it('should validate with ARIA rules as a svg element', async () => {
    expect(await axeComponent(Comp)).toHaveNoViolations()
  })
})

describe('Progress scss', () => {
  it('have to match snapshot', () => {
    const scss = loadScss(require.resolve('../style/dnb-progress.scss'))
    expect(scss).toMatchSnapshot()
  })
  it('have to match default theme snapshot', () => {
    const scss = loadScss(
      require.resolve('../style/themes/dnb-progress-theme-ui.scss')
    )
    expect(scss).toMatchSnapshot()
  })
})
