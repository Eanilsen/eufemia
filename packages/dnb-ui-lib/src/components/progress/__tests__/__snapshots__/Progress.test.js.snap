// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circular Progress component have to match snapshot 1`] = `
<Progress
  0={
    Object {
      "displayName": "Progress",
      "props": Object {
        "no_animation": "no_animation",
        "on_complete": [Function],
        "progress": "progress",
        "size": "'small'",
        "type": "'circular'",
        "visible": "visible",
      },
    }
  }
  no_animation={false}
  on_complete={null}
  progress={50}
  size="medium"
  type="circular"
  visible={true}
>
  <div
    className="dnb-progress dnb-progress--visible"
  >
    <ProgressCircular
      complete={false}
      maxOffset={88}
      progress={50}
      size="medium"
    >
      <div
        aria-label="50%"
        className="dnb-progress__circular dnb-progress__circular--medium dnb-progress__circular--has-progress"
        title="50%"
      >
        <Circle
          className="dnb-progress__circular__line light paused"
        >
          <svg
            className="dnb-progress__circular__line light paused"
            shapeRendering="geometricPrecision"
            viewBox="0 0 32 32"
          >
            <circle
              className="dnb-progress__circular__circle"
              cx="16"
              cy="16"
              fill="none"
              r="14"
              strokeWidth="4"
            />
          </svg>
        </Circle>
        <Circle
          className="dnb-progress__circular__line dark paused"
          style={
            Object {
              "strokeDashoffset": 44,
            }
          }
        >
          <svg
            className="dnb-progress__circular__line dark paused"
            shapeRendering="geometricPrecision"
            style={
              Object {
                "strokeDashoffset": 44,
              }
            }
            viewBox="0 0 32 32"
          >
            <circle
              className="dnb-progress__circular__circle"
              cx="16"
              cy="16"
              fill="none"
              r="14"
              strokeWidth="4"
            />
          </svg>
        </Circle>
      </div>
    </ProgressCircular>
  </div>
</Progress>
`;

exports[`Progress scss have to match default theme snapshot 1`] = `
"/*
* Progress theme
*
*/
/**
 * This file is only used to make themes independent
 * so that they can get imported individually, without the core styles
 *
 */
/*
 * Utilities
 */
.dnb-progress__circular__line {
  animation-duration: 2s; }

.dnb-progress__circular__line.light .dnb-progress__circular__circle {
  stroke: var(--color-mint-green); }

.dnb-progress__circular__line.dark .dnb-progress__circular__circle {
  stroke: var(--color-emerald-green); }
"
`;

exports[`Progress scss have to match snapshot 1`] = `
"/*
* DNB Progress
*
*/
/**
 * This file is only used to make components independent
 * so that they can get imported individually, without the core styles
 *
 */
/*
 * Utilities
 */
/*
 * Scopes
 *
 */
/*
 * Document Reset
 *
 */
.dnb-progress {
  font-family: var(--font-family-default);
  font-weight: var(--font-weight-default);
  font-size: 1rem;
  font-style: normal;
  line-height: 1.5rem;
  color: var(--color-black-80, #333);
  -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
  font-variant-numeric: lining-nums;
  font-feature-settings: \\"lnum\\";
  /**
  * 1. Remove repeating backgrounds in all browsers (opinionated).
  * 2. Add border box sizing in all browsers (opinionated).
  */
  /**
  * 1. Add text decoration inheritance in all browsers (opinionated).
  * 2. Add vertical alignment inheritance in all browsers (opinionated).
  */
  margin: 0;
  padding: 0; }
  .dnb-progress *,
  .dnb-progress ::before,
  .dnb-progress ::after {
    background-repeat: no-repeat;
    /* 1 */
    box-sizing: border-box;
    /* 2 */ }
  .dnb-progress ::before,
  .dnb-progress ::after {
    text-decoration: inherit;
    /* 1 */
    vertical-align: inherit;
    /* 2 */ }

/*
 * Progress component
 *
 */
:root {
  --progress-circular-circle: 88;
  --progress-circular-circle-offset--min: 88;
  --progress-circular-circle-offset--max: 1; }

.dnb-progress {
  position: relative; }
  .dnb-progress__circular {
    position: relative;
    width: 2rem;
    height: 2rem;
    transform: rotate(-90deg); }
    .dnb-progress__circular--small {
      width: 1rem;
      height: 1rem; }
    .dnb-progress__circular--large {
      width: 4rem;
      height: 4rem; }
    .dnb-progress__circular--huge {
      width: 20rem;
      height: 20rem; }
    .dnb-progress__circular svg {
      position: absolute;
      width: 100%;
      height: 100%; }
    .dnb-progress__circular__line {
      animation-duration: 2s;
      animation-delay: 200ms;
      animation-timing-function: linear;
      animation-iteration-count: infinite; }
    .dnb-progress__circular__line.background {
      stroke-dashoffset: var(--progress-circular-circle-offset--max); }
    .dnb-progress__circular__line.light {
      animation-name: progress-circular-line-light;
      stroke-dasharray: var(--progress-circular-circle), var(--progress-circular-circle);
      stroke-dashoffset: var(--progress-circular-circle-offset--max); }
    .dnb-progress__circular__line.dark {
      animation-name: progress-circular-line-dark;
      stroke-dasharray: var(--progress-circular-circle), var(--progress-circular-circle);
      stroke-dashoffset: var(--progress-circular-circle-offset--min); }
    .dnb-progress__circular__line.paused {
      animation-play-state: paused; }
    .dnb-progress__circular--has-progress .dnb-progress__circular__line.dark {
      transition: stroke-dashoffset 200ms ease-out; }
    .dnb-progress__circular__circle {
      stroke-linecap: round; }
    .dnb-progress__circular__line.light .dnb-progress__circular__circle {
      stroke: grey; }
    .dnb-progress__circular__line.dark .dnb-progress__circular__circle {
      stroke: black; }
  .dnb-progress:not(.dnb-progress--no-animation), .dnb-progress--visible:not(.dnb-progress--no-animation) {
    opacity: 0;
    animation: progress-fade-in 200ms ease-out 1 forwards; }
  .dnb-progress--complete:not(.dnb-progress--no-animation):not(.dnb-progress--visible) {
    animation: progress-fade-out 600ms ease-out 1 forwards; }

@keyframes progress-fade-in {
  0% {
    opacity: 0; }
  100% {
    opacity: 1; } }

@keyframes progress-fade-out {
  0% {
    opacity: 1; }
  100% {
    opacity: 0; } }

@keyframes progress-circular-line-light {
  0% {
    stroke-dashoffset: var(--progress-circular-circle-offset--min); }
  50% {
    stroke-dashoffset: var(--progress-circular-circle-offset--min); }
  100% {
    stroke-dashoffset: var(--progress-circular-circle-offset--max); } }

@keyframes progress-circular-line-dark {
  0% {
    stroke-dashoffset: var(--progress-circular-circle-offset--min); }
  50% {
    stroke-dashoffset: var(--progress-circular-circle-offset--max); }
  100% {
    stroke-dashoffset: var(--progress-circular-circle-offset--max); } }
"
`;
